lib/hpke/src/signature.cpp

```diff
@@ -95,69 +101,246 @@ struct ConcreteSignature : public Signature
+    , md(digest_to_md(digest))
+  {}
+
+  std::unique_ptr<Signature::PrivateKey> generate_key_pair() const override
+  {
+    throw std::runtime_error("Not implemented");
+  }
+
+  std::unique_ptr<Signature::PrivateKey> derive_key_pair(
+    const bytes& /*ikm*/) const override
+  {
+    throw std::runtime_error("Not implemented");
+  }
+
+  static std::unique_ptr<Signature::PrivateKey> generate_key_pair(size_t bits)
+  {
+    auto ctx = make_typed_unique(EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr));
+    if (ctx == nullptr) {
+      throw openssl_error();
+    }
+
+    if (EVP_PKEY_keygen_init(ctx.get()) <= 0) {
+      throw openssl_error();
+    }
+
+    // NOLINTNEXTLINE(hicpp-signed-bitwise)
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(ctx.get(), bits) <= 0) {
+      throw openssl_error();
+    }
+
+    auto* pkey = static_cast<EVP_PKEY*>(nullptr);
+    if (EVP_PKEY_keygen(ctx.get(), &pkey) <= 0) {
+      throw openssl_error();
+    }
+
+    return std::make_unique<PrivateKey>(pkey);
+  }
+
+  // TODO(rlb): Implement derive() with sizes
+
+  bytes serialize(const Signature::PublicKey& pk) const override
+  {
+    const auto& rpk = dynamic_cast<const PublicKey&>(pk);
+    int len = i2d_PublicKey(rpk.pkey.get(), nullptr);
+    auto raw = bytes(len);
+    auto* data_ptr = raw.data();
+    if (len != i2d_PublicKey(rpk.pkey.get(), &data_ptr)) {
+      throw openssl_error();
+    }
+    return raw;
+  }
+
+  std::unique_ptr<Signature::PublicKey> deserialize(
+    const bytes& enc) const override
+  {
+    const auto* data_ptr = enc.data();
+    auto* pkey = d2i_PublicKey(EVP_PKEY_RSA, nullptr, &data_ptr, enc.size());
+    if (pkey == nullptr) {
+      throw openssl_error();
+    }
+    return std::make_unique<RSASignature::PublicKey>(pkey);
+  }
+
+  bytes serialize_private(const Signature::PrivateKey& sk) const override
+  {
+    const auto& rsk = dynamic_cast<const PrivateKey&>(sk);
+    int len = i2d_PrivateKey(rsk.pkey.get(), nullptr);
+    auto raw = bytes(len);
+    auto* data_ptr = raw.data();
+    if (len != i2d_PrivateKey(rsk.pkey.get(), &data_ptr)) {
+      throw openssl_error();
+    }
+
+    return raw;
+  }
+
+  std::unique_ptr<Signature::PrivateKey> deserialize_private(
+    const bytes& skm) const override
+  {
+    const auto* data_ptr = skm.data();
+    auto* pkey = d2i_PrivateKey(EVP_PKEY_RSA, nullptr, &data_ptr, skm.size());
+    if (pkey == nullptr) {
+      throw openssl_error();
+    }
+    return std::make_unique<RSASignature::PrivateKey>(pkey);
+  }
+
+  bytes sign(const bytes& data, const Signature::PrivateKey& sk) const override
+  {
+    const auto& rsk = dynamic_cast<const PrivateKey&>(sk);
+
+    auto ctx = make_typed_unique(EVP_MD_CTX_create());
+    if (ctx == nullptr) {
+      throw openssl_error();
+    }
+
+    if (1 !=
+        EVP_DigestSignInit(ctx.get(), nullptr, md, nullptr, rsk.pkey.get())) {
+      throw openssl_error();
+    }
+
+    static const size_t max_sig_size = 256;
+    auto siglen = max_sig_size;
```

Comment:
```suggestion
    auto siglen = EVP_PKEY_size(rsk.pkey.get());
```

Also, please make the corresponding change for the group-based signatures.